packages/flutter/lib/src/cupertino/route.dart

```diff
@@ -378,56 +378,109 @@ class CupertinoPage<T> extends Page<T> {
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
   ///  * `primaryRouteAnimation` is a linear route animation from 0.0 to 1.0
   ///    when this screen is being pushed.
+  final Animation<double> primaryRouteAnimation;
+
   ///  * `secondaryRouteAnimation` is a linear route animation from 0.0 to 1.0
   ///    when another screen is being pushed on top of this one.
+  final Animation<double> secondaryRouteAnimation;
+
   ///  * `linearTransition` is whether to perform the transitions linearly.
   ///    Used to precisely track back gesture drags.
-  CupertinoPageTransition({
-    super.key,
-    required Animation<double> primaryRouteAnimation,
-    required Animation<double> secondaryRouteAnimation,
-    required this.child,
-    required bool linearTransition,
-  }) : _primaryPositionAnimation =
-           (linearTransition
-             ? primaryRouteAnimation
-             : CurvedAnimation(
-                 parent: primaryRouteAnimation,
+  final bool linearTransition;
+
+  @override
+  State<CupertinoPageTransition> createState() => _CupertinoPageTransitionState();
+}
+
+class _CupertinoPageTransitionState extends State<CupertinoPageTransition> {
+
+
+  // When this page is coming in to cover another page.
+  late Animation<Offset> _primaryPositionAnimation;
+  // When this page is becoming covered by another page.
+  late Animation<Offset> _secondaryPositionAnimation;
+  // Shadow of page which is coming in to cover another page.
+  late Animation<Decoration> _primaryShadowAnimation;
+  // Curve of primary page which is coming in to cover another page.
+  CurvedAnimation? _primaryPositionCurve;
+  // Curve of secondary page which is becoming covered by another page.
+  CurvedAnimation? _secondaryPositionCurve;
+  // Curve of primary page's shadow.
+  CurvedAnimation? _primaryShadowCurve;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupAnimation();
+  }
+
+  @override
+  void didUpdateWidget(covariant CupertinoPageTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
+    || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
+    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
+    _disposeCurve();
+    _setupAnimation();
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _disposeCurve();
+  }
+
+  void _disposeCurve() {
+    _primaryPositionCurve?.dispose();
+    _secondaryPositionCurve?.dispose();
+    _primaryShadowCurve?.dispose();
+  }
+
+  void _setupAnimation() {
+    _primaryPositionAnimation =
+           (widget.linearTransition
+             ? widget.primaryRouteAnimation
+             : _primaryPositionCurve = CurvedAnimation(
+                 parent: widget.primaryRouteAnimation,
                  curve: Curves.fastEaseInToSlowEaseOut,
                  reverseCurve: Curves.fastEaseInToSlowEaseOut.flipped,
```

Comment:
Careful, because here you might end up with disposing the curved animation twice. You are not resetting it to `null`